src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -279,16 +438,18 @@ ucs_status_t uct_tcp_sockcm_ep_recv(uct_tcp_sockcm_ep_t *cep)
         goto out;
     }
 
-    if (cep->comm_ctx.offset < sizeof(*hdr)) {
-        goto out;
-    }
+    if (!(cep->state & UCT_TCP_SOCKCM_EP_HDR_RECEIVED)) {
+        if (cep->comm_ctx.offset < sizeof(*hdr)) {
+            goto out;
+        }
 
-    hdr                  = (uct_tcp_sockcm_priv_data_hdr_t *)cep->comm_ctx.buf;
-    cep->comm_ctx.length = sizeof(*hdr) + hdr->length;
-    ucs_assertv(cep->comm_ctx.offset <= cep->comm_ctx.length , "%zu > %zu",
-                cep->comm_ctx.offset, cep->comm_ctx.length);
+        hdr                  = (uct_tcp_sockcm_priv_data_hdr_t *)cep->comm_ctx.buf;
+        cep->comm_ctx.length = sizeof(*hdr) + hdr->length;
+        ucs_assertv(cep->comm_ctx.offset <= cep->comm_ctx.length , "%zu > %zu",
+                    cep->comm_ctx.offset, cep->comm_ctx.length);
 
-    cep->state          |= UCT_TCP_SOCKCM_EP_RECEIVING;
+        cep->state          |= UCT_TCP_SOCKCM_EP_HDR_RECEIVED;
```

Comment:
but need to know how much will arrive. the header has this info and needs to be read only once.

Reply:
maybe check if length of data received >= sizeof(hdr)?

Reply:
this would be true every time after receiving the header but need to extract the length from it
(cep->comm_ctx.length = sizeof(*hdr) + hdr->length) 
just once only when the offset reaches the hdr size for the first time
(cep->comm_ctx.offset >= sizeof(*hdr)).

Reply:
ok